frontend/src/components/users/GridTools.ts

```diff
@@ -48,12 +47,80 @@ export const makeGridCoordinates = (widgets: Widgets) => {
   })
   return result
 }
-//위젯을 옮길 경우 차지하게 될 좌표 배열을 반환 [tools]
-// export const makeMoveCoordinates = (
-//   widgets: Widgets,
-//   index: number,
-//   coord: Coordinate
-// ) => {}
+
+//위젯을 옮길 경우 차지하게 될 좌표 배열을 반환 [완료][tools]
+export const makeMoveCoordinates = (widget: WidgetType, direction: Vec2) => {
+  return makePermutation(
+    widget.pos.add(direction),
+    widget.pos.add(widget.size).add(direction)
+  )
+}
+//위젯이 그리드 사이즈 범위 안에 있는지 확인해주는 함수 [완료][tools]
+const isInGridSize = (widget: WidgetType) => {
+  return (
+    widget.pos.v[0] >= 0 &&
+    widget.pos.v[1] >= 0 &&
+    widget.pos.v[0] + widget.size.v[0] - 1 < gridSize.v[0] &&
+    widget.pos.v[1] + widget.size.v[1] - 1 < gridSize.v[1]
+  )
+}
+
+//위젯을 밀 수 있는 지 확인하는 함수 [주기능]
+export const isPushable = (
+  widget: WidgetType,
+  cursorPosition: Vec2,
+  widgets: Widgets
+) => {
+  const tryPushWidgets: Widgets = Array.from(widgets) // 위젯들 밀면서 확인할 widgets의 사본
+  const movedRange = makeMoveCoordinates(widget, cursorPosition) //widget을 cursorPosition만큼 옮길 시 차지하는 좌표범위
+  const movedRangeWidgets = coordinateRangeWidgets(
+    widgets,
+    widget.pos.add(cursorPosition),
+    widget.pos.add(widget.size).add(cursorPosition)
+  )
+  const xList = movedRange.map(ele => ele.v[0])
+  const movedRangeMiddleX = (Math.max(...xList) + Math.min(...xList)) / 2
+
+  movedRangeWidgets.forEach(ele => {
+    if (ele !== widget) {
+      if (ele.pos.v[0] < movedRangeMiddleX) {
+        if (tryPush(ele, pos(-1, 0), tryPushWidgets) === true)
+          ele.pos = ele.pos.add(pos(-1, 0))
+        else if (tryPush(ele, pos(1, 0), tryPushWidgets) === true)
+          ele.pos = ele.pos.add(pos(1, 0))
+      } else {
+        if (tryPush(ele, pos(1, 0), tryPushWidgets) === true)
+          ele.pos = ele.pos.add(pos(1, 0))
+        else if (tryPush(ele, pos(-1, 0), tryPushWidgets) === true)
+          ele.pos = ele.pos.add(pos(-1, 0))
+      }
+    }
+  })
+
+  return movableToEmpty(widget, cursorPosition, tryPushWidgets) !== false
+}
+const tryPush = (widget: WidgetType, direction: Vec2, widgets: Widgets) => {
```

Comment:
```suggestion
const isPushableTo = (widget: WidgetType, direction: Vec2, widgets: Widgets) => {
```
특정 방향으로 옮길 수 있는지 여부를 확인하고, bool 값을 반환하므로 (`is ~`), `isPushableTo` 같은 이름이 좋을 것 같습니다.

Reply:
반영했습니다